src/objective-c/GRPCClient/GRPCCall.m

```diff
@@ -52,6 +55,165 @@ @interface GRPCCall ()<GRXWriteable>
 @property(atomic, strong) NSDictionary *responseHeaders;
 @property(atomic, strong) NSDictionary *responseTrailers;
 @property(atomic) BOOL isWaitingForToken;
+
+- (instancetype)initWithHost:(NSString *)host
+                        path:(NSString *)path
+                  callSafety:(GRPCCallSafety)safety
+              requestsWriter:(GRXWriter *)requestsWriter
+                 callOptions:(GRPCCallOptions *)callOptions;
+
+@end
+
+@implementation GRPCRequestOptions
+
+- (instancetype)initWithHost:(NSString *)host path:(NSString *)path safety:(GRPCCallSafety)safety {
+  if ((self = [super init])) {
+    _host = host;
+    _path = path;
+    _safety = safety;
+  }
+  return self;
+}
+
+- (id)copyWithZone:(NSZone *)zone {
+  GRPCRequestOptions *request =
+      [[GRPCRequestOptions alloc] initWithHost:[_host copy] path:[_path copy] safety:_safety];
+
+  return request;
+}
+
+@end
+
+@implementation GRPCCall2 {
+  GRPCCallOptions *_callOptions;
+  id<GRPCResponseHandler> _handler;
+
+  GRPCCall *_call;
+  BOOL _initialMetadataPublished;
+  GRXBufferedPipe *_pipe;
+  dispatch_queue_t _dispatchQueue;
+}
+
+- (instancetype)initWithRequestOptions:(GRPCRequestOptions *)requestOptions
+                               handler:(id<GRPCResponseHandler>)handler
+                           callOptions:(GRPCCallOptions *)callOptions {
+  if (!requestOptions || !requestOptions.host || !requestOptions.path) {
+    [NSException raise:NSInvalidArgumentException format:@"Neither host nor path can be nil."];
+  }
+
+  if ((self = [super init])) {
+    _requestOptions = [requestOptions copy];
+    _callOptions = [callOptions copy];
+    _handler = handler;
+    _initialMetadataPublished = NO;
+    _pipe = [GRXBufferedPipe pipe];
+    _dispatchQueue = dispatch_queue_create(NULL, DISPATCH_QUEUE_SERIAL);
+  }
+
+  return self;
+}
+
+- (instancetype)initWithRequestOptions:(GRPCRequestOptions *)requestOptions
+                               handler:(id<GRPCResponseHandler>)handler {
+  return [self initWithRequestOptions:requestOptions handler:handler callOptions:nil];
+}
+
+- (void)start {
+  dispatch_async(_dispatchQueue, ^{
+    if (!self->_callOptions) {
+      self->_callOptions = [[GRPCCallOptions alloc] init];
+    }
+
+    self->_call = [[GRPCCall alloc] initWithHost:self->_requestOptions.host
+                                            path:self->_requestOptions.path
+                                      callSafety:self->_requestOptions.safety
+                                  requestsWriter:self->_pipe
+                                     callOptions:self->_callOptions];
+    if (self->_callOptions.initialMetadata) {
+      [self->_call.requestHeaders addEntriesFromDictionary:self->_callOptions.initialMetadata];
+    }
+    id<GRXWriteable> responseWriteable = [[GRXWriteable alloc] initWithValueHandler:^(id value) {
+      dispatch_async(self->_dispatchQueue, ^{
```

Comment:
As a general comment, having two blocks here makes it a little hard to follow. Consider splitting the value handler and completion handler content into private helper methods.